cashsvr_input_miscta_params.c

```diff
@@ -0,0 +1,217 @@
+	if (!miscta_lnk.ta_read) {
+		ALOGE("%s: DLSYM failed for ta_read", __func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int cash_miscta_read_unit(uint32_t id, uint32_t *data_out)
+{
+	uint32_t ta_unit_sz = 0;
+	int rc;
+
+	rc = miscta_lnk.ta_getsize(id, &ta_unit_sz);
+	if (ta_unit_sz == 0) {
+		ALOGE("FATAL: Cannot get MiscTA unit %u size", id);
+		return -1;
+	};
+
+	rc = miscta_lnk.ta_read(id, data_out, ta_unit_sz);
+	if (rc) {
+		ALOGE("WARNING: MiscTA unit %u read returns %d", id, rc);
+	};
+
+	return 0;
+}
+
+/*
+ * cash_miscta_read_store_params - Read configuration from MiscTA and
+ *				   store it to the CASH data file
+ *
+ * Call this function with force == true to force params read and
+ * store even if the calibration file already exists.
+ */
+static int cash_miscta_read_store_params(bool force)
+{
+	struct cash_tamisc_calib_params tacfg;
+	struct stat st = {0};
+	bool read_error = false, set_perm = false;
+	int rc, fd, i;
+
+	/* Check if the file exists to eventually set the owner */
+	if (stat(CASHSERVER_CALDATA_FILE, &st) == -1) {
+		set_perm = true;
+	} else if (!force) {
+		/* If file exists and not forcing re-read, just exit. */
+		return 0;
+	}
+
+	fd = open(CASHSERVER_CALDATA_FILE, O_WRONLY | O_CREAT, 0664);
+	if (fd < 0) {
+		ALOGE("FATAL: Cannot open/create %s", CASHSERVER_CALDATA_FILE);
+		return -1;
+	} else if (set_perm) {
+		/* Set owner on the file, if it has been just created */
+		chown(CASHSERVER_CALDATA_FILE, AID_INPUT, AID_SYSTEM);
+	}
+
+	/* Wire up the function pointers */
+	rc = cash_miscta_wire_up();
+	if (rc)
+		return rc;
+
+	/* This can be painfully slow */
+	for (i = 0; i < 12; i++) {
+		rc = miscta_lnk.ta_open(2, 0x1, 1);
+		if (!rc)
+			break;
+
+		sleep(5);
+	}
+	if (rc) {
+		ALOGE("FATAL: Cannot open MiscTA");
+		return -1;
+	}
+
+	rc = 0;
+
+	rc = cash_miscta_read_unit(TA_UNIT_RGBCIR_CAPS1, &tacfg.rgbcir_caps1);
+	rc += cash_miscta_read_unit(TA_UNIT_RGBCIR_CAPS2, &tacfg.rgbcir_caps2);
+	if (rc) {
+		ALOGE("FATAL: Cannot read RGBCIR config from MiscTA\n");
+		read_error = true;
+	}
+
+	/* Number of SPADs */
+	rc = cash_miscta_read_unit(TA_UNIT_TOF_SPAD_NUM, &tacfg.tof_spad_num);
+
+	/* Type of SPADs: Aperture Type == 1 - Non-Aperture Type == 0 */
+	rc += cash_miscta_read_unit(TA_UNIT_TOF_SPAD_TYPE,
+					&tacfg.tof_spad_type);
+
+	/* Calibration Data offset, expressed in micrometers */
+	rc += cash_miscta_read_unit(TA_UNIT_TOF_UM_OFFSET,
+					&tacfg.tof_um_offset);
+	if (rc) {
+		ALOGE("FATAL: Cannot read ToF config from MiscTA\n");
+		ALOGE("Your ToF sensor may work suboptimally.\n");
+		read_error = true;
+	}
+
+	miscta_lnk.ta_close();
+
+	/* Write the struct into the calibration file */
+	write(fd, &tacfg, sizeof(struct cash_tamisc_calib_params));
+	close(fd);
+
+	if (read_error)
+		return -1;
+
+	return rc;
+}
+
+int cash_miscta_init_params(struct cash_tamisc_calib_params *conf)
+{
+	struct stat st = {0};
+	int ret, fd;
+
+	/* Create folder, if it doesn't exist */
```

Comment:
Let's leave that to the cashsvr.rc init file, because... let's just say selinux isn't happy.